p2p-sync/src/main/java/org/smartregister/p2p/presenter/P2PReceiverPresenter.java

```diff
@@ -165,65 +172,256 @@ public void onAuthorizationDetailsProvided(@NonNull Map<String, Object> authoriz
     public void onConnectionRejected(@NonNull String endpointId, @NonNull ConnectionResolution connectionResolution) {
         view.showToast(view.getString(R.string.receiver_rejected_the_connection), Toast.LENGTH_LONG);
         resetState();
-        startAdvertisingMode();
+        prepareForAdvertising(false);
     }
 
     @Override
     public void onConnectionUnknownError(@NonNull String endpointId, @NonNull ConnectionResolution connectionResolution) {
         //Todo: Go back to advertising mode
         //Todo: And show the user an error
-        view.showToast(view.getString(R.string.an_error_occurred_before_acceptance_or_rejection), Toast.LENGTH_LONG);
-        resetState();
-        startAdvertisingMode();
+        if (getCurrentPeerDevice() != null && endpointId.equals(getCurrentPeerDevice().getEndpointId())) {
+            view.showToast(view.getString(R.string.an_error_occurred_before_acceptance_or_rejection), Toast.LENGTH_LONG);
+            resetState();
+            prepareForAdvertising(false);
+        }
     }
 
     @Override
     public void onConnectionBroken(@NonNull String endpointId) {
         //Todo: Show the user an error
         //Todo: Go back to advertising mode
-        resetState();
-        view.showToast(String.format(view.getString(R.string.connection_to_endpoint_broken), endpointId), Toast.LENGTH_LONG);
-        startAdvertisingMode();
+        if (getCurrentPeerDevice() != null && endpointId.equals(getCurrentPeerDevice().getEndpointId())) {
```

Comment:
Shouldn't there be an else

Reply:
The else is not supposed to be possible, if it is, then due to the memory leak whose issue is here https://github.com/OpenSRP/android-p2p-sync/issues/39. 

The app was crashing without the if since the code was being run even after exiting the activity and the ConnectionsClient is holding on to the callbacks passed

 If the [google issue](https://issuetracker.google.com/issues/129429944) has no reply by the time I am done with the remaining issues and no other solution in site, then we'll have to use WeakReferences for the view (Or clean it up) but the API might still hold onto the listener until it makes the last call

in the else, means that the callback was called although we are not connected to the endpoint or have already disconnected